sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/RocksDBSuite.scala

```diff
@@ -1736,6 +1737,84 @@ class RocksDBSuite extends AlsoTestWithChangelogCheckpointingEnabled with Shared
     }
   }
 
+  testWithChangelogCheckpointingEnabled("load the same version of pending snapshot uploading") {
+    val remoteDir = Utils.createTempDir().toString
+    val conf = dbConf.copy(minDeltasForSnapshot = 2, compactOnCommit = false)
+    new File(remoteDir).delete() // to make sure that the directory gets created
+    withDB(remoteDir, conf = conf) { db =>
+      db.load(0)
+      db.put("foo", "bar")
+      db.commit()
+
+      db.load(1)
+      db.put("foo", "bar")
+      db.commit()
+      db.doMaintenance()
+
+      db.load(1)
+      db.put("foo", "bar")
+      db.commit()
+
+      db.load(2)
+      db.put("foo", "bar")
+      db.commit()
+
+      db.load(2)
+      db.put("foo", "bar")
+      db.commit()
+
+      db.doMaintenance()
+
+      db.load(2)
+      db.put("foo", "bar")
+      db.commit()
+    }
+  }
+
+  for (random_seed <- 1 to 16) {
+    testWithChangelogCheckpointingEnabled(s"randomized snapshotting $random_seed") {
+      val remoteDir = Utils.createTempDir().toString
+      val conf = dbConf.copy(minDeltasForSnapshot = 3, compactOnCommit = false)
+      new File(remoteDir).delete() // to make sure that the directory gets created
+      withDB(remoteDir, conf = conf) { db =>
+        val random = new Random(random_seed)
+        var curVer: Int = 0
+        for (i <- 1 to 100) {
+          db.load(curVer)
+          db.put("foo", "bar")
+          db.commit()
+          if (random.nextInt(5) == 0) {
```

Comment:
Why do we choose `5` and `2` for this test ? Should we also print some additional logs that might help to know which current version we are processing ?

Reply:
Sure. I can comment on why 5 and 2, but the 2 is to have a balance between moving on to the next version and keep to the current one. 5 is to simulate skip checkpoints and make it more likely to delay to more than one snapshot scheduling, considering snapshot checkpointing is every 3 batches.